### ISSUE TITLE
BotSession.close() logs ERROR java.net.SocketException: Socket closed

### ISSUE THREAD

**Issue by ceefour** [rubenlagus/TelegramBots#168]
Currently it's only possible to register a handler, but not to unregister a handler.

For https://github.com/fast4indonesia/yoopabot , I need to be able to dynamically add/edit/remove handlers without restarting the JVM.

**Comment by rubenlagus** [comment]
Do you mean for LongPolling or Webhook bots?

**Comment by ceefour** [comment]
For LongPolling bots

**Comment by rubenlagus** [comment]
If you check the code, when registering a longpolling bot, the api is making nothing but create and return BotSession object. That session have convenient methods to start/stop at wish.

**Comment by ceefour** [comment]
After calling `BotSession.close()` it logs Exception:

**Comment by rubenlagus** [comment]
That socket exception could be raised due to different causes (intentional closing among them). If I change it to INFO, it may cause that it is hidden in other important cases. In any case, I'm open to suggestions or pull requests.

**Comment by rubenlagus** [comment]
@ceefour any improvement in latest versions? Any suggestion?

**Comment by mr-yusupov** [comment]
Hello @rubenlagus 
Is there any news on this issue? Still getting the same exception in CDI context.

> BOTSESSION: java.net.SocketException: Socket closed

**Comment by rubenlagus** [comment]
There is no actual problem with this exception when called during "close" request since we are actually forcing it close regardless its state. The only option is to reduce logging level but that can hide it in other cases where important.

**Comment by myplacedk** [comment]
This problem seems to be reduced to a race condition since opening this issue.

It seems to be a symptom of imperfect thread handling. I tried cleaning it up a bit, and that seemed to solve it. Although I'm not sure if it will affect performance negatively when under heavy load.

The problem was that the session could be closed while `getUpdatesFromServer()` was running. One workaround is to check `running` again just before `httpclient.execute(httpPost)`, but I solved it by synchronizing more code.
